src/h3lib/lib/polygon.c

```diff
@@ -93,3 +95,172 @@ bool pointInsidePolygon(const GeoPolygon* geoPolygon, const BBox* bboxes,
 #undef IS_EMPTY
 #undef INIT_ITERATION
 #undef ITERATE
+
+/**
+ * Count the number of polygons containing a given loop.
+ * TODO: It would be really nice to memoize this.
+ * @param  loop         Loop to count containers for
+ * @param  polygons     Polygons to test
+ * @param  polygonCount Number of polygons in the test array
+ * @return              Number of polygons containing the loop
+ */
+static int countContainers(LinkedGeoLoop* loop, LinkedGeoPolygon** polygons,
+                           BBox** bboxes, int polygonCount) {
+    int containerCount = 0;
+    for (int i = 0; i < polygonCount; i++) {
+        if (pointInsideLinkedGeoLoop(polygons[i]->first, bboxes[i],
+                                     &loop->first->vertex)) {
+            containerCount++;
+        }
+    }
+    return containerCount;
+}
+
+/**
+ * Find the polygon to which a given hole should be allocated. Note that this
+ * function will return null if no parent is found.
+ * @param  loop         Inner loop describing a hole
+ * @param  polygon      Head of a linked list of polygons to check
+ * @param  polygonCount Number of polygons to check
+ * @return              Pointer to parent polygon, or null if not found
+ */
+static LinkedGeoPolygon* findPolygonForHole(LinkedGeoLoop* loop,
+                                            LinkedGeoPolygon* polygon,
+                                            BBox* bboxes, int polygonCount) {
+    // Early exit with no polygons
+    if (polygonCount == 0) {
+        return NULL;
+    }
+    // Initialize arrays for candidate loops and their bounding boxes
+    LinkedGeoPolygon** candidates =
+        calloc(polygonCount, sizeof(LinkedGeoPolygon*));
+    assert(candidates != NULL);
+    BBox** candidateBBoxes = calloc(polygonCount, sizeof(BBox*));
+    assert(candidateBBoxes != NULL);
+
+    // Find all polygons that contain the loop
+    int containerCount = 0;
+    int index = 0;
+    while (polygon) {
+        // We are guaranteed not to overlap, so just test the first point
+        if (pointInsideLinkedGeoLoop(polygon->first, &bboxes[index],
+                                     &loop->first->vertex)) {
+            candidates[containerCount] = polygon;
+            candidateBBoxes[containerCount] = &bboxes[index];
+            containerCount++;
+        }
+        polygon = polygon->next;
+        index++;
+    }
+
+    // Set the initial return value to the first candidate
+    LinkedGeoPolygon* parent = candidates[0];
```

Comment:
Probably better as a `const LinkedGeoPolygon*` seeing as you do not intend to change it.

Reply:
I *do* intend to change it - this is the current candidate, but gets overridden on L168 if there's another one with more parents.

Reply:
Understood. Thanks for the effort.